File: cloud_alm_claude.py
import os
from typing import Any, Dict, Optional

import anthropic
import requests
from dotenv import load_dotenv

load_dotenv()

class ClaudeClient:
    def __init__(self, api_key: Optional[str] = None, model: Optional[str] = None):
        self.api_key = api_key or os.getenv("ANTHROPIC_API_KEY")
        self.model = model or os.getenv("ANTHROPIC_MODEL", "claude-3.5")
        if not self.api_key:
            raise ValueError("ANTHROPIC_API_KEY is required")
        self.client = anthropic.Client(api_key=self.api_key)

    def summarize(self, text: str, context: str = "") -> str:
        prompt = (
            anthropic.HUMAN_PROMPT
            + "You are a SAP Cloud ALM integration assistant. Generate a concise summary of the following work item details and highlight any recommended actions."
            + "\n\nContext:\n"
            + context
            + "\n\nWork item content:\n"
            + text
            + "\n\nSummary:"
            + anthropic.AI_PROMPT
        )
        response = self.client.completions.create(
            model=self.model,
            prompt=prompt,
            max_tokens_to_sample=400,
            temperature=0.3,
        )
        return response.completion.strip()

    def generate_issue_description(self, title: str, details: str) -> str:
        prompt = (
            anthropic.HUMAN_PROMPT
            + "You are a SAP Cloud ALM integration assistant. Create a clear, professional issue description that can be posted into SAP Cloud ALM."
            + "\n\nTitle:\n"
            + title
            + "\n\nDetails:\n"
            + details
            + "\n\nIssue description:"
            + anthropic.AI_PROMPT
        )
        response = self.client.completions.create(
            model=self.model,
            prompt=prompt,
            max_tokens_to_sample=400,
            temperature=0.4,
        )
        return response.completion.strip()


class CloudALMClient:
    def __init__(self, base_url: Optional[str] = None, api_token: Optional[str] = None):
        self.base_url = base_url or os.getenv("SAP_CLOUD_ALM_API_BASE_URL")
        self.api_token = api_token or os.getenv("SAP_CLOUD_ALM_API_TOKEN")
        if not self.base_url or not self.api_token:
            raise ValueError("SAP_CLOUD_ALM_API_BASE_URL and SAP_CLOUD_ALM_API_TOKEN are required")

    def _headers(self) -> Dict[str, str]:
        return {
            "Authorization": f"Bearer {self.api_token}",
            "Content-Type": "application/json",
        }

    def fetch_work_item(self, work_item_id: str) -> Dict[str, Any]:
        url = f"{self.base_url.rstrip('/')}/workitems/{work_item_id}"
        response = requests.get(url, headers=self._headers())
        response.raise_for_status()
        return response.json()

    def create_comment(self, work_item_id: str, comment: str) -> Dict[str, Any]:
        url = f"{self.base_url.rstrip('/')}/workitems/{work_item_id}/comments"
        payload = {"text": comment}
        response = requests.post(url, headers=self._headers(), json=payload)
        response.raise_for_status()
        return response.json()

    def create_work_item(self, title: str, description: str) -> Dict[str, Any]:
        url = f"{self.base_url.rstrip('/')}/workitems"
        payload = {"title": title, "description": description}
        response = requests.post(url, headers=self._headers(), json=payload)
        response.raise_for_status()
        return response.json()


def summarize_cloud_alm_work_item(work_item_id: str) -> str:
    cloud_alm = CloudALMClient()
    claude = ClaudeClient()
    work_item = cloud_alm.fetch_work_item(work_item_id)

    title = work_item.get("title", "")
    description = work_item.get("description", "")
    details = f"Title: {title}\nDescription: {description}"

    summary = claude.summarize(details, context="SAP Cloud ALM work item")
    return summary

File: src/cloudAlmClaude.js
import axios from "axios";
import dotenv from "dotenv";

dotenv.config();

const ANTHROPIC_API_KEY = process.env.ANTHROPIC_API_KEY;
const ANTHROPIC_MODEL = process.env.ANTHROPIC_MODEL || "claude-3.5";
const SAP_CLOUD_ALM_API_BASE_URL = process.env.SAP_CLOUD_ALM_API_BASE_URL;
const SAP_CLOUD_ALM_API_TOKEN = process.env.SAP_CLOUD_ALM_API_TOKEN;

if (!ANTHROPIC_API_KEY) {
  throw new Error("ANTHROPIC_API_KEY is required in .env");
}

if (!SAP_CLOUD_ALM_API_BASE_URL || !SAP_CLOUD_ALM_API_TOKEN) {
  throw new Error("SAP_CLOUD_ALM_API_BASE_URL and SAP_CLOUD_ALM_API_TOKEN are required in .env");
}

const anthropicClient = axios.create({
  baseURL: "https://api.anthropic.com/v1",
  headers: {
    "x-api-key": ANTHROPIC_API_KEY,
    "Content-Type": "application/json",
  },
});

const cloudAlmClient = axios.create({
  baseURL: SAP_CLOUD_ALM_API_BASE_URL.replace(/\/$/, ""),
  headers: {
    Authorization: `Bearer ${SAP_CLOUD_ALM_API_TOKEN}`,
    "Content-Type": "application/json",
  },
});

export async function summarizeText(text, context = "") {
  const prompt = `\nHuman: You are a SAP Cloud ALM integration assistant. Generate a concise summary of the following work item details and highlight any recommended actions.\n\nContext:\n${context}\n\nWork item content:\n${text}\n\nAssistant:`;
  const response = await anthropicClient.post("/complete", {
    model: ANTHROPIC_MODEL,
    prompt,
    max_tokens_to_sample: 400,
    temperature: 0.3,
  });
  return response.data?.completion?.trim() ?? "";
}

export async function generateIssueDescription(title, details) {
  const prompt = `\nHuman: You are a SAP Cloud ALM integration assistant. Create a clear, professional issue description that can be posted into SAP Cloud ALM.\n\nTitle:\n${title}\n\nDetails:\n${details}\n\nAssistant:`;
  const response = await anthropicClient.post("/complete", {
    model: ANTHROPIC_MODEL,
    prompt,
    max_tokens_to_sample: 400,
    temperature: 0.4,
  });
  return response.data?.completion?.trim() ?? "";
}

export async function fetchWorkItem(workItemId) {
  const response = await cloudAlmClient.get(`/workitems/${encodeURIComponent(workItemId)}`);
  return response.data;
}

export async function createComment(workItemId, comment) {
  const response = await cloudAlmClient.post(
    `/workitems/${encodeURIComponent(workItemId)}/comments`,
    { text: comment }
  );
  return response.data;
}

export async function createWorkItem(title, description) {
  const response = await cloudAlmClient.post("/workitems", {
    title,
    description,
  });
  return response.data;
}
